### PR TITLE
perf(sizeshape): scatter-based spatial moments + inertia, replacing regionprops einsum

### PR DESCRIPTION
**Pull request by timtreis**
Replace regionprops' per-region einsum for the `moments` / `moments_central` / `moments_normalized` / `moments_hu` / `inertia_tensor` columns with a single label-scatter pass (`primitives/_moments.py`), and derive axis lengths / eccentricity / orientation from the scatter central moments (option B) so `get_sizeshape` requests no regionprops moments at all (~1.6×). Matches regionprops to round-off (raw moments bit-exact); the eigenvalue clip keeps thin objects NaN-free like skimage.

Assumes the contiguous 1..N contract: segment index is `label - 1` and per-object origins come from one `scipy.ndimage.find_objects` pass — no lookup table (this rebase drops the `label_to_idx_lut` / `primitives/segment.py` dependency that #89 removed).

### CODE REVIEW

File: .github/scripts/run_benchmark.sh [hunk 1]
```diff
@@ -4,6 +4,11 @@
 # Usage: run_benchmark.sh <out-dir> <head-commit-sha>
 set -euo pipefail
 
+# Pin BLAS/OpenMP to one thread so timings reflect algorithmic cost, not incidental parallelism
+# (cp_measure keeps core functions single-threaded; the batch layer is what parallelises).
+export OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 MKL_NUM_THREADS=1
```
Comment (afermg) [inline .github/scripts/run_benchmark.sh:9]: Doesn't this mean that the current CI may have incidental parallelism?

File: src/cp_measure/core/measureobjectsizeshape.py [hunk 2]
```diff
@@ -7,6 +7,12 @@
 import scipy.ndimage
 import skimage.measure
 from centrosome import zernike
+from cp_measure.primitives._moments import (
```
Comment (afermg) [inline src/cp_measure/core/measureobjectsizeshape.py:10]: Some of the calculations (axes, inertia, moments) were replaced by faster internal implementations

File: test/test_sizeshape_moments.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
+"""Equivalence tests for the scatter-based spatial moments in ``get_sizeshape``.
+
+These lock **numerical fidelity to skimage regionprops**: the scatter ``spatial_moments_2d`` /
```
Comment (afermg) [inline test/test_sizeshape_moments.py:3]: Ok, so numerical fidelity wrt skimage, makes sense to me

File: src/cp_measure/primitives/_moments.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    NDArray[numpy.floating],
+    NDArray[numpy.floating],
+]:
+    """Per-object ``(raw, central, normalized, hu)`` spatial moments for a 2D label image.
+
+    ``raw`` / ``central`` / ``normalized`` are ``(n, 4, 4)`` and ``hu`` is ``(n, 7)``, ordered by
+    ascending label. Drop-in for the matching ``regionprops_table`` columns
+    (``moments-p-q`` etc.); ``normalized`` is NaN where ``p + q < 2`` (skimage convention).
+    """
+    n = int(labels.max())
```
Comment (afermg) [inline src/cp_measure/primitives/_moments.py]: Do we need to cast to integers? n is only used for equivalence to 0, and AFAIK it `numpy.int(0) == 0`

File: src/cp_measure/primitives/_moments.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    if n == 0:
+        empty = numpy.zeros((0, _ORDER, _ORDER))
+        return empty, empty, empty, numpy.zeros((0, 7))
+
+    # Contiguous 1..N (see cp_measure._sanitize): find_objects index i is label i+1. Stack each
```
Comment (afermg) [inline src/cp_measure/primitives/_moments.py:71]: This reproduces the usual trick of "tile-ifying" the masks. It's interesting how widespread that trick holds speedups.

File: test/test_sizeshape_moments.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        masks,
+        properties=[
+            "axis_major_length",
+            "axis_minor_length",
+            "eccentricity",
+            "orientation",
+        ],
+    )
+    numpy.testing.assert_allclose(
+        out["MajorAxisLength"], ref["axis_major_length"], rtol=1e-9, atol=1e-9
```
Comment (afermg) [inline test/test_sizeshape_moments.py]: Could you clarify why we are not using the global tolerance level set at the top of the test file? Same for the examples below

File: test/test_sizeshape_moments.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    )
+
+
+def test_axes_match_single_object():
+    masks = numpy.zeros((64, 64), numpy.int32)
+    masks[18:50, 20:45] = 1  # rectangle -> nonzero orientation, distinct axis lengths
+    _assert_axes_match(masks)
+
+
+# Exact output column order of the PyPI 0.1.19 release (2D, new_features + calculate_advanced on).
```
Comment (afermg) [inline test/test_sizeshape_moments.py]: Unless I'm missing something, the schema order shouldn't matter (we can use a hash to compare values), what does matter (and would break things is if we are missing features or have extra, except for the "new features"). I think this test should check that instead of the explicit order.

File: test/test_sizeshape_moments.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+
+def test_sizeshape_key_order_matches_release():
+    masks = numpy.zeros((40, 40), numpy.int32)
+    masks[5:25, 5:25] = 1
+    assert list(get_sizeshape(masks, masks.astype(float))) == _RELEASE_KEY_ORDER
+
+
+def test_axes_clip_thin_objects_match_skimage():
+    # Thin / oblique objects have a (near-)singular inertia tensor; float error can drive the minor
```
Comment (afermg) [inline test/test_sizeshape_moments.py]: I don't understand what "near-singular" means in this context.

File: test/test_sizeshape_moments.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+    # eigenvalue slightly negative. inertia_2d clips to 0 like skimage, so axis lengths and
+    # eccentricity match regionprops and are never NaN. Pre-clip, ~4% of these gave NaN axis_minor
+    # / eccentricity > 1 — this loop guards that regression.
```
Comment (afermg) [inline test/test_sizeshape_moments.py]: Hmm, perhaps it is a bit excessive to test this, but I think it makes sense since we are ensuring that the original skimage clipping behaviour is maintained.

File: test/test_sizeshape_moments.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+        out = get_sizeshape(masks, masks.astype(float))
+        assert not numpy.isnan(out["MinorAxisLength"]).any()
+        assert not numpy.isnan(out["Eccentricity"]).any()
+        ref = skimage.measure.regionprops_table(
+            masks, properties=["axis_minor_length", "eccentricity"]
+        )
+        # atol=1e-6: a 1px line's minor axis is float noise (~4e-8) in skimage vs exactly 0
+        # from our clip; both mean "zero width", so compare only above that floor.
+        numpy.testing.assert_allclose(
+            out["MinorAxisLength"], ref["axis_minor_length"], rtol=1e-7, atol=1e-6
```
Comment (afermg) [inline test/test_sizeshape_moments.py]: Same as before, aren't we using a global tolerance?

File: test/test_sizeshape_moments.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        )
+        numpy.testing.assert_allclose(
+            out["Eccentricity"], ref["eccentricity"], rtol=1e-7, atol=1e-6
+        )
+
+
+def test_axes_match_single_pixel_and_line():
```
Comment (afermg) [inline test/test_sizeshape_moments.py:295]: edge cases, these are worth testing.

**Review by afermg — CHANGES_REQUESTED** [review]
I requested some minor changes, mostly justifying the presence of global tolerance level and adjusting one to not assume a defined feature order. Once those are sorted out it should be an easy approval.

File: .github/scripts/run_benchmark.sh (hunk 1 above)
Comment (timtreis) [reply .github/scripts/run_benchmark.sh:9]: Potentially yeah, depending on how the CI runners were configured & set up

File: src/cp_measure/core/measureobjectsizeshape.py (hunk 2 above)
Comment (timtreis) [reply src/cp_measure/core/measureobjectsizeshape.py:10]: Yeah, the goal was to remove the per-region moment einsum calls

File: test/test_sizeshape_moments.py (hunk 8 above)
Comment (timtreis) [reply test/test_sizeshape_moments.py]: reworded to "a thin / 1px-wide object has almost no width, so the smaller eigenvalue of its inertia tensor is ~0; float round-off can push it slightly below zero"

File: test/test_sizeshape_moments.py (hunk 7 above)
Comment (timtreis) [reply test/test_sizeshape_moments.py]: Changed to set-logic

File: test/test_sizeshape_moments.py (hunk 9 above)
Comment (timtreis) [reply test/test_sizeshape_moments.py]: Yeah 🤔

**Review by afermg — APPROVED** [review]
Propagating the global tolerances seems like the right choice. I'm still not fully convinced about the new test (test_moments_match_disk), but I think it's simple enough that it doesn't greatly increase complexity.

**Comment by timtreis** [comment]
The test was to actually test a situation in which scaling can land "inbetween pixels" 👌
